Core changes for `k6/x/execution` JS module

imiric opened the pull request:
These are core changes needed for the [`k6/x/execution` JS module](https://github.com/k6io/xk6-execution).

Initially the module was part of core as well, but we decided to extract it into an extension to be able to quickly iterate on the API. See the JS usage and example in the extension repo.

Part of #1320

imiric commented on js/runner_test.go:
```diff
@@ -2001,3 +2001,98 @@ func TestComplicatedFileImportsForGRPC(t *testing.T) {
+			if (ts.vus_active !== 1) throw new Error('unexpected vus_active: '+ts.vus_active);
+			if (ts.vus_max !== 0) throw new Error('unexpected vus_max: '+ts.vus_max);
+			if (ts.iterations_completed !== 0) throw new Error('unexpected iterations_completed: '+ts.iterations_completed);
+			if (ts.iterations_interrupted !== 0) throw new Error('unexpected iterations_interrupted: '+ts.iterations_interrupted);
```
These checks are kind of pointless... Should I move these tests to some higher-level ones that involve actual executors?

MStoykov commented on js/modules/k6/execution/execution.go:
```diff
@@ -0,0 +1,94 @@
+		return nil, errors.New("scenario information can only be returned from an exported function")
+	}
+
+	progress, _ := ss.ProgressFn()
+	out := map[string]interface{}{
+		"name":      ss.Name,
+		"executor":  ss.Executor,
+		"startTime": ss.StartTime,
+		"progress":  progress,
+	}
```
I think it will be better if `progress` calls the method when accessed (or is a function) because:
1. this won't call `progressFn` unless the user actually needs it instead of needing the `name` for example
2. will allow users to monitor the progress without calling `getScenarioStats` constantly

imiric replied:
Yeah, it would be a slight performance benefit to evaluate it on access. I'll look into doing this with `DynamicObject` then.

MStoykov replied:
I would say that a normal Object with `defineProperty` will be fine, but a `DynamicObject` will do the job as well ... just a little too much IMO

imiric replied:
Resolved by c034057e?

MStoykov commented on js/modules/k6/execution/execution.go:
```diff
@@ -0,0 +1,94 @@
+func (e *Execution) GetTestStats(ctx context.Context) (map[string]interface{}, error) {
+	es := lib.GetExecutionState(ctx)
+	if es == nil {
+		return nil, errors.New("test information can only be returned from an exported function")
```
The message is ... not worded very well and not consistent with the wording in other places where we talk about the `init context` .

imiric replied:
True, I'll make it consistent.

imiric replied:
Resolved by 327a370a.

MStoykov reviewed (commented):
Haven't taken a deeper look or tried the code, but some comments:
1. IMO we should allow getting the info in all cases where any info is available
2. The things that call functions should probably call the functions on the spot (without any caching) or be functions in the first place
3. Given that a lot of the `*State` objects seem "static" can we have 1 of those and just give it everywhere
4. given `3.` (and even without it) can you check and make certain that a user script can't change any of the values?

imiric commented:
@MStoykov:

> IMO we should allow getting the info in all cases where any info is available

Are you referring to being able to call this from init? Most of this information doesn't make sense outside the context of a running VU/scenario. Only `vus_max` seems relevant in init and we could get it from the config instead of `*ExecutionState`, but besides complicating things slightly it would be weird to have to document which information is available and which isn't, instead of just disallowing it entirely. If we go with your 2nd suggestion below, it might make sense, so I'll think about it some more. :)

> The things that call functions should probably call the functions on the spot (without any caching) or be functions in the first place

Hhmm I'm not sure I prefer the API to be something like `getTestStats().duration()` or complicating it with `DynamicObject` like we discussed in #1320. Returning an already evaluated object is easier to understand (users don't expect proxies) and these are small objects so the performance benefit would likely be negligible.

I could give it a shot if you feel strongly about it, though would prefer the `DynamicObject` approach in that case.

> Given that a lot of the `*State` objects seem "static" can we have 1 of those and just give it everywhere

Hhmm you mean about `*ScenarioState` specifically and how it's set in the context each time a VU is activated? True, it would be better if it was set on the context without being passed around in `VUActivationParams`. I'll look into it, thanks.

But `*State` and `*ExecutionState` are only set once.

> given `3.` (and even without it) can you check and make certain that a user script can't change any of the values?

Sure. With dumb / already evaluated objects this shouldn't matter, but I'll confirm after your suggested changes.

na-- commented:
I have only skimmed the code changes, so I haven't fully grokked the PR yet, but I have a more generic comment here. I think the MVP version of https://github.com/loadimpact/k6/issues/1320 shouldn't be exposing information users can already mostly get. It should be exposing things that they cannot currently get in any way. Primarily because it would require some refactoring to the executors and may change how the rest of the feature is implemented.

From what I can see, `exec.getVUStats()` would return exactly `__VU` and `__ITER`, right? While it is _nice_ to have a better UX than the current super-global "constants" for that information, it's not _necessary_ to do so initially, users manage just fine without it. `exec.getScenarioStats()` is new information, so better, but even now you can get most of that information by just having different `exec` function (or via environment variables in each scenario). The rest of the information from `exec.getScenarioStats()` and the whole `exec.getTestStats()` is novel and previously inaccessible information, true, but probably wouldn't actually be useful to users from inside of the script... 

I think the MVP of this feature should have these two pieces of information, with everything else being a nice-to-have bonus:
- _in the current scenario_, what is the ID number of the VU currently running the iteration
- _in the current scenario_, what is the sequential number of the iteration for this VU that is currently being executed
- for `shared-iterations`, `constant-arrival-rate` and `variable-arrival-rate` scenarios, a way to get the sequential number of the current iteration among _all_ of the iterations for the current scenario, across all VUs and even instances.

The first 2 are the equivalent of `__VU` and `__ITER`, but only in terms of a single scenario. The last one is a very useful feature for data sharding and it should be easy to calculate with the execution segments. But all of these are things are information that's completely unavailable to user scripts right now. So, yeah, I may be wrong, but I think we should start with the "hard" (i.e. unknown) parts of this issue first, and implement the "easy" (i.e. known) parts last, as a polish to everything.

imiric commented:
@na-- OK, so to clarify, that would involve adding a few additional executor-specific counters, right? Except for the last global across instances values which could be calculated given the execution segment?

na-- commented:
> OK, so to clarify, that would involve adding a few additional executor-specific counters, right?

Yeah, basically a local VU number for the scenario (basically immutable after the VU activation in the scenario) and a local iteration number for the specific VU in the scenario (mutable, but the only lock contention here would be between whatever increments it and this new `exec` method used to retrieve it, so pretty light). Still, as I said, those things require some refactoring of the executors... :disappointed:

imiric commented:
So it's unfortunately safe to say that this won't make the v0.31.0 release. :disappointed:

I think the functionality described [here](https://github.com/loadimpact/k6/pull/1863#issuecomment-782069940) is implemented now, but it will need thorough testing and actually making the tests pass consistently.

Please give the code an overview, the commits should be fairly isolated, and test it on your machine to ensure it's at least functionally OK.

What's pending:

- Fixing the flaky tests. I think I managed to find a timing sweet spot for the `*GlobalIters` tests (they passed for more than 30m with [this script](https://dave.cheney.net/2013/06/19/stress-test-your-go-packages)), but it's clear from the Windows failures that the others are still flaky. I'm not sure how to refactor them to avoid this, so any ideas are welcome.

- Addressing @MStoykov's comments above, which I started doing but ran into issues with sharing the object across goja runtimes.

- Addressing any additional feedback.

imiric commented on lib/executor/helpers.go:
```diff
@@ -108,6 +108,7 @@ func getIterationRunner(
 
 			// TODO: move emission of end-of-iteration metrics here?
 			executionState.AddFullIterations(1)
+			incrIter()
```
We're currently only tracking completed/full iterations. Would tracking interrupted iterations too be interesting to users? It would essentially double the amount of counters for everything... :-/

imiric commented:
@MStoykov @na-- Could you take another look and let me know if there's anything pending here?

From what we discussed, allowing the use of this module in the init context is the major thing still missing. After taking another look, I still think it would require substantial refactoring to achieve, and most of the information just doesn't make sense in the init context.

The big (only?) use case that does make sense is retrieving the max test duration before the test starts, as mentioned on the [forum](https://community.k6.io/t/get-total-test-duration/1567/2) and [here](https://github.com/k6io/k6/issues/1320#issuecomment-812463418). The problem with this is that the init context is evaluated before the `ExecutionScheduler`/`ExecutionState` is created.
This is a chicken/egg situation since we need to have access to the consolidated options, including the ones exported in JS, to be able to return the correct duration, but these aren't consolidated until after the init context is evaluated for the first time (see `cmd/run.go`).
I considered adding `lib.Options` to `common.InitEnvironment` to precalculate the max duration/VUs and avoid depending on `ExecutionScheduler`, but we run into the same chicken/egg thing, since options aren't consolidated when `InitEnvironment` is first initialized.

A possible approach would be to offer a pure JS API to calculate and return just the max duration, but that would duplicate a lot of the same logic we have in Go, and it doesn't seem worth it considering it can be done with the current implementation from the `default` function, and once we have #1001 to abort the test, it would fulfill the requirement of the person on the forum.

MStoykov commented on lib/execution_segment.go:
```diff
@@ -730,3 +731,96 @@ func (et *ExecutionTuple) GetNewExecutionTupleFromValue(value int64) (*Execution
+}
+
+// GoTo sets the scaled index to its biggest value for which the corresponding
+// unscaled index is is smaller or equal to value.
```
```suggestion
// unscaled index is smaller or equal to value.
```

MStoykov commented on lib/execution_segment.go:
```diff
@@ -730,3 +731,96 @@ func (et *ExecutionTuple) GetNewExecutionTupleFromValue(value int64) (*Execution
 	}, nil
 }
+
+// SegmentedIndex ...
```
maybe write something here ;)

imiric replied:
:sweat_smile: I was hoping one of you guys could come up with a useful description for this, I wasn't sure how to document it.

imiric replied:
@na-- Care to contribute a useful description for this? :smile:

na-- replied:
How about this?
```suggestion
// SegmentedIndex is an iterator that returns both the scaled and the unscaled 
// sequential values according to the given ExecutionTuple. It is not thread-safe,
// concurrent access has to be externally synchronized.
```

MStoykov commented on lib/executor/constant_arrival_rate.go:
```diff
@@ -196,9 +199,24 @@ func (car *ConstantArrivalRate) Init(ctx context.Context) error {
+// incrGlobalIter increments the global iteration count for this executor,
+// taking into account the configured execution segment.
+func (car *ConstantArrivalRate) incrGlobalIter() {
+	car.segIdx.Next()
+	atomic.StoreUint64(car.globalIter, uint64(car.segIdx.GetUnscaled()))
```
This is logically racy as there is nothing preventing 2 goroutines to call `car.segIdx.Next()` and then `atomic.StoreUint64` and the one who got the "first" values from `segIdx.Next` to store it second, effectively having the previous value in `car.globalIter`.

Same for the `shared-iterations`

imiric replied:
Let me know if it's as simple as 8973a841.

I wanted to remove the use of `atomic`, but we still need it because `globalIter` needs to remain a pointer since `incrGlobalIter` is passed to the VU `State` which is where it's being called (this is part of the refactor to fix how `globalIter` works).

imiric commented on js/runner_test.go:
```diff
@@ -2006,3 +2006,101 @@ func TestComplicatedFileImportsForGRPC(t *testing.T) {
+	testCases := []struct {
+		name, script, expErr string
+	}{
+		{"vu_ok", `
```
Not sure I understand the drawbacks, but I don't like using maps for test cases because of unordered test runs. I know it shouldn't matter and tests shouldn't rely on the order, but as long as they don't I prefer knowing the order they start in.

MStoykov replied:
> I don't like using maps for test cases because of unordered test runs.

I am not saying to use maps but use `{name: "vu_ok", script: "some javascript"}`

>I know it shouldn't matter and tests shouldn't rely on the order, but as long as they don't I prefer knowing the order they start in.

Also https://go-review.googlesource.com/c/go/+/310033/ and .. if it doesn't matter it's better to be random so at least it continues to not matter ;)

MStoykov replied:
Also `t.Parallel` in practice should mean that they are already somewhat randomized, the proposed `-shuffle` shuffles top-level tests only from the looks of it ;)

imiric replied:
> I am not saying to use maps but use `{name: "vu_ok", script: "some javascript"}`

Ah, I see what you mean now :) Sure, I'll change it.

imiric replied:
Resolved by 8b24539a.

imiric commented:
@MStoykov Let me know if the recent refactor, specifically 69e67506, addresses your feedback about how scenario iterations work. They're now being incremented and set in the VU `State`, which keeps them stable within a single iteration. I also moved the regular VU iteration increment to the same place, and now `__ITER` is no longer set for `setup()` and `teardown()`.

Can you please confirm with the test script you posted on Slack?

There are still a couple of tests pending (test ideas are welcome!) that I'll add later today, and the existing ones are possibly flaky (especially `TestSharedIterationsGlobalIters`) but let me know if there's something glaringly wrong with the current state.

imiric commented on lib/state.go:
```diff
@@ -67,8 +67,22 @@ type State struct {
+func (s *State) Init() {
+	s.scenarioVUID = make(map[string]uint64)
+	s.scenarioVUIter = make(map[string]int64)
+	s.iteration, s.scIter, s.scIterGlobal = -1, -1, -1
```
This is... nasty. I'd much prefer it if we could keep these as `uint64`, but since the increment needs to happen before `runFn()`, this seemed like the path of least resistance that doesn't involve a lot of additional checks.

Any better ideas?

MStoykov replied:
I guess they can be `^0` and adding 1 will overflow them ? does that have some problems 
https://play.golang.org/p/71mrOn9ATxg
you need the cast as otherwise the magical constant magic in go doesn't translate it to uint64 before `^` and instead it things it is the int `-1` :facepalm:

imiric replied:
Hhmm could be, I didn't think of that :smile: :+1:

Is relying on int overflow a valid pattern in Go?

imiric replied:
Resolved by 6fc76f85.

BTW @MStoykov is there a reason you used `^(uint64)(0)` instead of `^uint64(0)`?

MStoykov commented on lib/state.go:
```diff
@@ -79,3 +95,67 @@ func (s *State) CloneTags() map[string]string {
+	} else {
+		s.scenarioVUIter[s.ScenarioName] = 0
+	}
+	if s.IncrScIter != nil {
+		s.scIter = s.IncrScIter()
+	}
+	if s.IncrScIterGlobal != nil {
+		s.scIterGlobal = s.IncrScIterGlobal()
+	}
+}
```
This is also logically racy as there is nothing stopping a goroutine going to `IncrScIter` and then before it gets to IncrScIterGlobal 20 others to get to it which leads to this 
```

INFO[0006] 8 9 6 1                                       source=console
INFO[0006] 10 11 5 1                                     source=console
INFO[0006] 12 13 2 1                                     source=console
INFO[0006] 14 15 5 2                                     source=console
INFO[0006] 18 19 5 3                                     source=console
INFO[0006] 17 18 11 0                                    source=console
INFO[0006] 11 12 1 1                                     source=console
INFO[0006] 16 17 4 1                                     source=console
...
```
from 
```
    var stats = execution.getScenarioStats()
    if (stats.iterationGlobal != stats.iteration) {
    console.log(stats.iterationGlobal, stats.iteration, __VU, __ITER)
    }
```

MStoykov replied:
This seems like a bad idea idea either way as it lets anything having access to `lib.State` to update the iteration (which was also possible before to be honest, but I still don't like it)
IMO :
1.  `s.iteration` should be part of  InitializedVU 
2. scIter and globalIter be part of ActivatedVU 

And both to somehow get queried from there, but to be update from their respected places *when* an iteration gets started.

Given that the whole iterations things are mostly a roundabout way to do #1539 and that something like [xk6-segment](https://github.com/MStoykov/xk6-segment) will be way more useful (not requiring to have only 1 item per iteration) I am fine with all the iteration counters to be dropped at this point as they seem to be problematic and we can get  1 PR without any of them and then you to do another with what we can do as long as both you and @na-- agree.

imiric replied:
Hhmm you're right. I don't know what I was thinking with the `State.scIterMx` mutex, it wouldn't work across VUs :man_facepalming:

I see your point about incrementing iterations from `State` being problematic, since it's part of the context and widely shared. Moving the fields to `InitializedVU`/`ActivatedVU` does make sense.

I'll look into fixing this, but I wouldn't want to drop anything at this point, since we've delayed this feature long enough. Let's discuss it later this week.

MStoykov replied:
I am mostly proposing that we scale this back to without the iterations so we can merge it more easily and then you to make another PR ... with just the more ... involved iteration changes

imiric replied:
@MStoykov Please take a look at 9cc9de4b. There are some wonky things about it still, like the setting of lambdas on `lib.State` that are passed via `VUActivationParams`, but I think this sync issue should be fixed now.

We can only retrieve iterations via `State` now, which I needed to expose to `k6/execution`, but the increment is done internally in `js.ActiveVU`.

MStoykov reviewed (commented):
I tried to do partly the proposed changes, but hit some problems :( so I don't know how reasonable it is.

Also I noticed that a linter was complaining that `SharedIterations.Run` is copying a lock/mutex due to the method being with a non pointer receiver. So if we are going to have lock/mutexes embedded we need all methods (which will be using it but probably better to just be *all*) to be with pointer receivers.

I haven't really looked at most of the other code as those problems seemed big enough to me to look into :(

MStoykov commented on lib/state.go:
```diff
@@ -67,8 +68,23 @@ type State struct {
+
+// Init initializes some private state fields.
+func (s *State) Init() {
+	s.scenarioVUID = make(map[string]uint64)
```
This seems to me like it can be not a map ;)

as in reality it gets set whenever you activate the VU and then we get it through using the scenarioName so it can just as well be a single uint64

imiric replied:
I'm not sure it can though. We set it only once if it's not defined for a particular scenario, and `js.VU` is never destroyed so these IDs can persist after being shared among scenarios. There's no need to overwrite it on each VU activation, as the VU ID should remain static throughout the test run.

MStoykov commented on lib/state.go:
```diff
@@ -67,8 +68,23 @@ type State struct {
+// Init initializes some private state fields.
+func (s *State) Init() {
+	s.scenarioVUID = make(map[string]uint64)
+	s.scenarioVUIter = make(map[string]int64)
```
I see why this is map ... I do think it is needed in the current way things are done, but I hope you can drop it with the proposed refactoring as I truly dislike it :sweat_smile:

imiric replied:
Why? :smile: I think it makes sense to persist this across VU activations for each scenario, otherwise it would get messy with VU sharing between scenarios.

If this is the same comment as [the previous one](https://github.com/k6io/k6/pull/1863#discussion_r643780977), let's resolve this and discuss it there.

MStoykov commented on js/modules/k6/execution/execution.go:
```diff
@@ -0,0 +1,170 @@
+	stats := map[string]interface{}{
+		// XXX: For consistency, should this be startTime instead, or startTime
+		// in ScenarioStats be converted to duration?
+		"duration": func() goja.Value {
+			return rt.ToValue(es.GetCurrentTestRunDuration().String())
+		},
```
I think this might be better off as `progress` although I also would expect people would want the max duration of the test which might be put here as `duration` ;)

imiric replied:
Hhmm so what's your suggestion? :smile:

Rename this to `progress` and convert it to float, and add a (max)`duration` property?

na-- replied:
I think `duration` is fine (and we might add `maxDuration` in the future, not now). Though I'd like it to be a float number in milliseconds instead of a string, so JS people don't have to parse stringy times.

imiric replied:
Resolved by 7f420594.

MStoykov reviewed (commented):
I left some comments while doing a second pass on mostly not the iteration stuff:
LGTM but I need to either see it without the iteration changes or with the fixed iterations changes for better comments :(, but I don't see some big problems with it :+1:

imiric commented:
@MStoykov

> Also I noticed that a linter was complaining that SharedIterations.Run is copying a lock/mutex due to the method being with a non pointer receiver.

Which linter? I haven't noticed this. I think there would be failures in the `*GlobalIters` tests if the `iterMx` mutex is not locking properly (i.e. each VU got a different mutex instance). Since `getNextGlobalIter` is defined on a pointer, I would assume the same instance is being used. I'll confirm this tomorrow, but you're saying that all `Run` methods should be defined on a pointer receiver?

MStoykov commented:
@imiric 
```
$ golangci-lint run lib/executor | grep copylocks
lib/executor/shared_iterations.go:208:10: copylocks: Run passes lock by value: go.k6.io/k6/lib/executor.SharedIterations contains sync.Mutex (govet)
```

I am saying that now that there is a `sync.Mutex` probably all methods need to be with pointer receivers as otherwise it might be very confusing in the future when someone uses it ... and it doesn't work in the `Run` method as it is on a non-pointer receiver and it gets copied ;)

MStoykov commented on js/runner.go:
```diff
@@ -743,6 +771,31 @@ func (u *VU) runFn(
 }
 
+func (u *ActiveVU) incrIteration() {
+	u.iteration++
+	u.state.Iteration = u.iteration
+
+	if u.iterSync != nil {
+		// block other VUs from incrementing scenario iterations
+		u.iterSync <- struct{}{}
+		defer func() {
+			<-u.iterSync // unlock
+		}()
+	}
+
+	if _, ok := u.scenarioIter[u.scenarioName]; ok {
+		u.scenarioIter[u.scenarioName]++
+	} else {
+		u.scenarioIter[u.scenarioName] = 0
+	}
+	if u.getNextScLocalIter != nil {
+		u.scIterLocal = u.getNextScLocalIter()
+	}
+	if u.getNextScGlobalIter != nil {
+		u.scIterGlobal = u.getNextScGlobalIter()
+	}
+}
+
```
Why do we not have GlobalIter for some executors?

I do think whenever we have local one we can (and should) have a global one and all executors do have 
1. way to make SegmetnedIndex
2. make iterations somehow 
which just means it needs to call `next` and get the scaled and unscaled values of it ... preferably atomically as I di in [xk6-segment](https://github.com/MStoykov/xk6-segment/blob/8df18d74cb838a7c16167afec6feabb65ae93af0/pkg/segment/segment.go#L112-L122). Even in golang you don't need the additional result type ... it is only useful for javascript.

imiric replied:
> Why do we not have GlobalIter for some executors?

I think I asked the same question to @na-- since [he mentioned these 3 explicitly in the specs](https://github.com/k6io/k6/pull/1863#issuecomment-782069940), and IIRC he doesn't think it would be useful for the other executors. I do agree that it would make things neater if we could just define this on `BaseExecutor`, but honestly I don't mind the current state much.

I don't quite understand what would be the benefit of changing `next`. The only difference I can see in the `xk6-segment` implementation is that we would avoid the additional read lock by not having to call `GetUnscaled()`. But semantically it makes more sense to me for `Next()` to only increment and to have a separate getter. It's also more flexible to use and I would have to change how it's used in `ramping-vus`.

MStoykov replied:
I would like to get @na-- to review this PR before I continue commenting after this PR. I think I am pulling this PR in a different direction than what you think it should be and would like to get a third opinion if I or you are right on the different issues or if ... we are both wrong :rofl: 

Regarding this ... I think you keep slapping synchronization on top of things trying to fix them not synchronizing , while I am trying to see if we can do with the amount of synchronization we have currently. Some thoughts:

- SegmentedIndex doesn't need to be thread-safe for `ramping-vus`, so it means that its performance will suffer (somewhat) for something it doesn't need
- The places where SegementedIndex needs to be thread-safe are the same places where it needs to be threadsafe from `next`/`prev`/`GoTo` to `GetScaled` and `GetUnscaled` (might've gotten a name wrong, I ain't looking at the code at the moment).  For me, this means that there should be `SyncSegmentedIndex`(better name exist) that returns the `scaled`, `unscaled` in the cases where it needs to. Even if we only care about `next` we can probably just have an even smaller wrapper for now.
- Again you already lock for `next` and then ... you lock once more around it (through additional means) to get the 2 counters you care about, so combined with the below I think you can have it all done at once with *1* synchronization mechanism wrapping a nonsynchronized SegmentedIndex which can still be used in the cases where synchronization ... isn't required.
- I see no reason why we shouldn't have iteration counters for all executors ... we might want them only for specific ones, but there is nothing wrong (and it definitely will be useful) for even the externally-controlled one to have them so they can do the same tricks the `shared-iterations` can do. And given that the current implementation clearly will be simpler if all of them have it instead of only some - I am for all of them having it.

na-- replied:
> > Why do we not have GlobalIter for some executors?
> 
> I think I asked the same question to @na-- since [he mentioned these 3 explicitly in the specs](https://github.com/k6io/k6/pull/1863#issuecomment-782069940), and IIRC he doesn't think it would be useful for the other executors.

I am currently reviewing this PR and still have a lot of things I need to understand in it, but if you're referencing this sentence of mine:
> for `shared-iterations`, `constant-arrival-rate` and `variable-arrival-rate` scenarios, a way to get the sequential number of the current iteration among _all_ of the iterations for the current scenario, across all VUs and even instances.

Then the reason for listing only these 3 executors is simple - they are the only ones where we can actually _know_ the sequential number of an iteration globally, among all VUs and instances, without any synchronization. This is because the iteration number is the thing we segment on, not VUs, as is the case with the rest of the executors.

na-- replied:
I am still reviewing this (getting easily distracted and finding it a bit difficult to make headway :disappointed: ), but yeah, as @MStoykov mentioned above, something in the synchronization looks a bit wrong or unnecessary.

I think his first point is about the fact that `SegmentedIndex.Next()` has locking: https://github.com/k6io/k6/blob/25ee787380f7917ca33115711612a6a6d724f29f/lib/execution_segment.go#L750-L758
But we also lock where it's used, for example: https://github.com/k6io/k6/blob/25ee787380f7917ca33115711612a6a6d724f29f/lib/executor/shared_iterations.go#L193-L197

If `Next()` (and the other traversal methods) returned the new scaled and unscaled values, you wouldn't need to lock twice and ` GetScaled()` and `GetUnscaled()` will also be obsolete, I think?

MStoykov replied:
yes, that is what I meant.

imiric replied:
OK, I'll try refactoring this then.

imiric replied:
Take a look at e9feff31. I decided to keep returning a struct instead of multiple values, to simplify future expansion.

imiric replied:
The other comment here is about adding global iterations to all executors. @na-- WDYT?

MStoykov replied:
https://github.com/k6io/k6/pull/2057 what I actually meant ... I think've missed initially that you have the local iteration as a counter that you atomically increment. 

The commit doesn't fix any comments and the names are not great, but I think that + the commit message should be good enough explanation

imiric commented on lib/state.go:
```diff
@@ -67,8 +67,20 @@ type State struct {
+	// These will be assigned on VU activation.
+	// Returns the iteration number of this VU in the current scenario.
+	GetScenarioVUIter func() uint64
+	// Returns the iteration number across all VUs in the current scenario
+	// unique to this single k6 instance.
+	// TODO: Maybe this doesn't belong here but in ScenarioState?
+	GetScenarioLocalVUIter func() uint64
+	// Returns the iteration number across all VUs in the current scenario
+	// unique globally across k6 instances (taking into account execution
+	// segments).
+	GetScenarioGlobalVUIter func() uint64
```
These 2 aren't relevant to a single VU/State, but I added them here since it was simpler to call them from the same place where we increment the other iterations. They would probably be better off as part of `ScenarioState` instead, but that would imply exposing a `ScenarioState.IncrIteration()` method, which we wanted to avoid for the VU `State`, so it's the same situation.

This doesn't bother me too much, but let me know if it should be changed.

imiric reviewed (commented):
@MStoykov I fixed the `copylocks` warning in 4a57d75e, but I changed `sync.Mutex` to a pointer instead of `Run`, to keep the change localized to this issue.

MStoykov commented on lib/executor/constant_arrival_rate.go:
```diff
@@ -196,9 +197,22 @@ func (car *ConstantArrivalRate) Init(ctx context.Context) error {
+// iteration number will be unique across k6 instances.
+func (car *ConstantArrivalRate) getNextGlobalIter() uint64 {
+	res := car.segIdx.Next()
+	// iterations are 0-based
```
```suggestion
	// iterations are indexed from 0
```
nit: this sounds better to me

na-- commented on lib/execution_segment.go:
```diff
@@ -730,3 +730,79 @@ func (et *ExecutionTuple) GetNewExecutionTupleFromValue(value int64) (*Execution
+
+// NewSegmentedIndex returns a pointer to a new SegmentedIndex instance,
+// given a starting index, LCD and offsets as returned by GetStripedOffsets().
+func NewSegmentedIndex(start, lcd int64, offsets []int64) *SegmentedIndex {
```
btw, writing the comment suggestion above, we can probably make this constructor accept the `ExecutionTuple` directly. As far as I can see, this is the only way it's always going to be called, so we can just have it call `et.GetStripedOffsets()` internally?

imiric replied:
Done in ed07691a.

MStoykov commented on core/local/local_test.go:
```diff
@@ -1387,3 +1388,281 @@ func TestNewExecutionSchedulerHasWork(t *testing.T) {
+		};
+
+		export function cvus() {
+			const stats = Object.assign({scenario: 'cvus', globalVUID: __VU}, exec.getVUStats());
```
by globalVUID do you mean "global for the instance" or for the test (in case of execution segments usage). As it is definitely not the latter and making it will be a breaking change I don't think is worth it.

On the other hand we have no way to get the latter ("global VUI id for the whole test") which is something that is a particular problem for cloud tests.

We can probably add it similar to how global Iteration was added, but maybe in a separate PR?

imiric replied:
Maybe "global" was the wrong name for it, but it's `__VU`, so global for the test and instance, since no execution segments are used. It's just a label to ensure VUs are being reused in the test, but I can change it to something else if you want.

I don't think we discussed global VU IDs like you mention, so yeah, let's add it to a separate issue and PR.

na-- replied:
Should be fairly simple to add, so I'd prefer now if possible

na-- replied:
hmm actually it won't be all that simple, when you consider dynamically allocated VUs for the arrival-rate executors :disappointed: and if we currently don't calculate the total max VUs in every k6 instance, regardless of the segment, it might have performance implications if we just add this now... so probably a separate PR will be better :disappointed:

MStoykov replied:
> performance implications if we just add this now.

We do now calculate through segment indexes which number an iteration is globally ... so I doubt calculating that for a VU will be noticeable

na-- replied:
Tto calculate which number a VU is, you first need to get the total max number of possible VUs (i.e. without an execution segments), which may be a very expensive operation for some `ramping-vus` configs. Whereas the global iteration counter is just a segmented index with the current execution segment, for which we've already done the calculations :man_shrugging:

MStoykov replied:
why? 

as far as I am concern it is literally when you make a VU you have segmented index ... which gives it's local (unscaled) and global VU (scaled) number and that's it.

I don't care what is the max VUs it is the responsibility of the code making the VUs to make certain we don't go above that

na-- replied:
:man_facepalming: yeah, you are completely right, I had a stupid idea in my head about how this should be done, probably influenced by some cloud-related implementation details :sweat_smile: the simple segmented index will be perfectly adequate way to implement this :+1:

na-- commented on js/runner.go:
```diff
@@ -540,6 +542,10 @@ type VU struct {
 	setupData goja.Value
 
 	state *lib.State
+	// ID of this VU in each scenario
+	scenarioID map[string]uint64
+	// count of iterations executed by this VU in each scenario
+	scenarioIter map[string]uint64
```
Are these maps just because of the `ramping-vus` executor?

imiric replied:
Yeah, I thought it would be simpler to just keep the count for all scenarios in cases where VUs are released and reused, though it would be more efficient to just use `uint64` here instead. Mihail made the same comment above. I'll see if I can change it.

imiric replied:
Actually, taking a look at this again, I'm not sure if changing it would be better. The memory needed for these maps is negligible (a few entries at most?), but we need some state to know if a VU ID was previously assigned by that same executor:
  https://github.com/k6io/k6/blob/f9e0ab6550c21477dbdfcb6cd49be8d755a47f11/js/runner.go#L620-L622

This is what drives the `BaseExecutor.VUIDLocal` counter (which could probably be unexported), as otherwise `GetNextScVUID()` would have to be called from elsewhere. Maybe in each executor after `GetPlannedVU()`:
  https://github.com/k6io/k6/blob/f9e0ab6550c21477dbdfcb6cd49be8d755a47f11/lib/executor/shared_iterations.go#L277

But that would mean several implementations of it, when using a map is simpler and we wouldn't need to think about this for new executors that might reuse VUs.

Let me know if there's another reason both you and @MStoykov dislike this approach and well, any suggestions. :)

imiric commented on lib/executor/base_executor.go:
```diff
@@ -66,6 +83,12 @@ func (bs BaseExecutor) GetConfig() lib.ExecutorConfig {
 	return bs.config
 }
 
+// getNextLocalVUID increments and returns the next VU ID that's specific for
+// this executor (i.e. not global like __VU).
+func (bs BaseExecutor) getNextLocalVUID() uint64 {
+	return atomic.AddUint64(bs.VUIDLocal, 1)
+}
```
Is this still needed? Right now we have 3 VU ID values: local to a single scenario (this one), global to a single instance (`__VU`),  and global across instances (added in #2064). I gathered this one from [this comment](https://github.com/k6io/k6/pull/1863#issuecomment-782069940) ("_in the current scenario_, what is the ID number of the VU currently running the iteration"), but I'm wondering how useful that is in practice...

MStoykov replied:
I feel like the use case for this VU ID is if you have *multiple* scenarios and need to segment something based on the VUs between them. 

I would argue in most cases the *per scenario* iteration will be sufficient and probably all other cases are some kind of memory optimization that likely is better handled by SharedArray and co.

So I am slightly :+1: for dropping this particular counter, we can always put it back in the future. But I am by no means blocking the PR based on that,

imiric replied:
Honestly, I would prefer it if we switched to GUIDs instead of incremental IDs, and just have those instead of 3 different types. It would break certain use cases like "run this only on the _first_ VU", but IMO those are workarounds for other features we need like events/hooks. GUIDs would work in both local and distributed contexts, and we would avoid all these counters.

It would be a huge breaking change, but something to think about for v1+?

na-- replied:
hmm there's nothing stopping you from having GUIDs even now, just have this in the init context:
```js
const vuID = getMyFancyUUID();
```

Integer IDs, especially when we can have them globally unique and mostly sequential so easily, have other benefits that UUIDs don't have, especially when it comes to ad-hoc partitioning of data or work :man_shrugging: so I don't really see a downside in having them here

MStoykov commented on cmd/cloud.go:
```diff
@@ -251,14 +251,16 @@ This will execute the test on the k6 cloud service. Use "k6 login cloud" to auth
-			go func() {
-				showProgress(progressCtx, conf, []*pb.ProgressBar{progressBar}, logger)
-				progressBarWG.Done()
-			}()
+			progressBarWG := &sync.WaitGroup{}
+			if !quiet {
+				progressBarWG.Add(1)
+				defer progressBarWG.Wait()
+				go func() {
+					showProgress(progressCtx, conf, []*pb.ProgressBar{progressBar}, logger)
+					progressBarWG.Done()
+				}()
+			}
```
The problem is that:
1. `--quiet` just makes it way more likely (and possible at all) to see the race, as writing to stdout is *synchronous* so it basically makes the synchronization for us .... or just slows k6 down enough that it doesn't happen :man_shrugging: 
2. The problem actually (as somewhat, but not very well(IMO) shown by the stack trace) is that [GetProgress](https://github.com/k6io/k6/blob/master/lib/executor/base_executor.go#L75) is defined on the * non-pointer receiver* which means that when it's invoked it will *copy* the struct effectively *read*ing all of its fields. At the same time the `Init` *write*s to one of the fields which makes the race listed above.

So we should probably have all the `BaseExecutor` methods be on the pointer receiver as it won't copy the struct.

imiric replied:
Resolved by f9e0ab65. Thanks! :clap:

na-- commented on lib/executor/shared_iterations.go:
```diff
@@ -232,15 +234,26 @@ func (si SharedIterations) Run(parentCtx context.Context, out chan<- stats.Sampl
 	regDurationDone := regDurationCtx.Done()
 	runIteration := getIterationRunner(si.executionState, si.logger)
 
+	maxDurationCtx = lib.WithScenarioState(maxDurationCtx, &lib.ScenarioState{
```
I don't like how we attach this only to `maxDurationCtx`. Even if this works now, I think this can become a subtle source of bugs in the future, see https://play.golang.org/p/xAEuY_FOOC3

In essence, right now this is the diagram of the contexts currently looks:
```
parentCtx -> maxDurationCtx -> regDurationCtx
                           \-> maxDurationCtxWithScenarioState -> vuCtx
```

This is not necessarily a problem, even before the `vuCtx` was branched as a "sibling" to the `regDurationCtx`, but if we can move the scenario state attachment before the `maxDurationCtx` and `regDurationCtx` it seems to me that it will much safer. Something like this:
```
parentCtx -> parentCtxWithScenarioState -> maxDurationCtx -> regDurationCtx
                                                         \-> vuCtx
```

imiric replied:
Hhmm would something like [d5ffc781](https://github.com/imiric/k6/commit/d5ffc78183602b87c679aad0854ab2c454efa853) be an improvement?

The problem with attaching it to `parentCtx` is that we need to pass the `progressFn` function to `ScenarioState`. Since this is a lambda created after `getDurationContexts()` is called, I had to split the initialization of `ScenarioState` like this, which is not nice.

A better approach might be to expand the `lib.Executor` interface and refactor all executors to have a standalone `GetProgress()` method (renaming the current one to `GetProgressBar()`), and then attach `ScenarioState` before `Run()` is called:
https://github.com/k6io/k6/blob/f9e0ab6550c21477dbdfcb6cd49be8d755a47f11/core/local/local.go#L329

But that would involve a much larger refactor to avoid what is a hypothetical source of bugs, so I don't want to start that here.

na-- replied:
Yeah, https://github.com/imiric/k6/commit/d5ffc78183602b87c679aad0854ab2c454efa853 has other issues that might also be a similarly subtle source of future bugs :sweat_smile: And I don't really have a better idea how to not have them without a big executor refactor like the one you propose ... :disappointed:  We could probably have some of this and `getDurationContexts()` wrapped in a new helper function, but the progress will still have to be "dangling" and attached separately.

In any case, we can ignore this for now, but I'll add a new issue to maybe refactor this in the future.

na-- replied:
Created https://github.com/k6io/k6/issues/2068, resolving this thread so it doesn't clutter the review.

na-- commented on js/modules/k6/execution/execution.go:
```diff
@@ -0,0 +1,169 @@
+	stats := map[string]interface{}{
+		"id":         vuState.VUID,
+		"idGlobal":   vuState.VUIDGlobal,
+		"idScenario": vuState.VUIDScenario,
+		"iteration":  vuState.Iteration,
+		"iterationScenario": func() goja.Value {
```
```suggestion
		"idInScenario": vuState.VUIDScenario,
		"iteration":  vuState.Iteration,
		"iterationInScenario": func() goja.Value {
```

This sounds just a bit better to me, but it's a minor nitpick. @MStoykov @imiric WDYT?

imiric replied:
Sure, I'm fine with that. Just that `idGlobal` and `iterationGlobal` in `getScenarioStats()` would slightly stick out, and renaming them to something like `idInTest` would be a bit more confusing. We'll have to document each of these properly anyway, so having them shorter and following the same naming convention would be better UX. We'd also document them in the TypeScript definitions, so the name itself doesn't matter much.

na-- commented on lib/executor/base_executor.go:
```diff
@@ -38,53 +40,74 @@ import (
+// getNextLocalVUID increments and returns the next VU ID that's specific for
+// this executor (i.e. not global like __VU).
+func (bs *BaseExecutor) getNextLocalVUID() uint64 {
+	return atomic.AddUint64(bs.VUIDLocal, 1)
```
Shouldn't this also be segmented? Right now with it being per-instance, I am not sure how useful it will be?

imiric replied:
:-/ We brought up actually removing this counter in [this thread](https://github.com/k6io/k6/pull/1863#discussion_r651867771). I'm not sure how helpful it is already, but even less so if we add another global variant of it. I think we already have a lot of possibly confusing values and adding anything new without knowing if users will actually need it only forces us to support it on the off chance someone is depending on it. The module is already large enough, so I'd rather leave this as is or remove it and add it later if needed.

na-- replied:
Ah, I replied in that previous thread only to your latest message, without reading the start... :sweat_smile: :+1: for removing this counter and potentially adding it back in the future. I could argue why a segmented per-scenario VU index in would be useful, but we can certainly add it later.

na-- replied:
Hmm actually, a segmented index here would be most useful for things like `constant-vus` and `ramping-vus` and maybe `per-vu-iterations`, where the VU number is the leading thing, not the iteration number, so that's what people would likely want to segment on :man_shrugging: 

So, yeah, we can remove it for now, so we can ship the PR more quickly, but I'm almost certain we're going to have to put a segmented per-scenario VU index back fairly soon :sweat_smile: Especially since we still don't have a per-scenario VU init function (https://github.com/k6io/k6/issues/785) where something like this could be generated easily by users.

imiric replied:
OK, I removed it in 670b7126 :disappointed: FYI @MStoykov

na-- commented on js/modules/k6/execution/execution.go:
```diff
@@ -0,0 +1,169 @@
+	stats := map[string]interface{}{
+		"name":      ss.Name,
+		"executor":  ss.Executor,
+		"startTime": ss.StartTime,
```
This probably needs a type conversion, either to a proper string or to some sort of a unix timestamp (with milliseconds). Right now, it's Go's `time.Time` type and behaves very strangely in JS. For example:
```js
export default function () {
    let scenarioStats = execution.getScenarioStats();
    console.log(scenarioStats.startTime);
    console.log(JSON.stringify(scenarioStats.startTime));
    console.log(typeof scenarioStats.startTime);
    console.log(scenarioStats.startTime instanceof Date);
    console.log(typeof scenarioStats.startTime.getMonth);
    console.log(Object.keys(scenarioStats.startTime));
}
```

will result in this:
```
2021-06-17 16:06:17.00787938 +0300 EEST m=+0.488634128
"2021-06-17T16:06:17.00787938+03:00"
object
false
undefined
add,addDate,after,appendFormat,before,clock,date,day,equal,format,gobEncode,hour,iSOWeek,in,isZero,local,location,marshalBinary,marshalJSON,marshalText,minute,month,nanosecond,round,second,string,sub,truncate,uTC,unix,unixNano,weekday,year,yearDay,zone
```

imiric replied:
Yeah, I ran into this when I was writing the tests, see https://github.com/dop251/goja/issues/170.

I'm now converting it in Go, see 80671dfa.

MStoykov replied:
hm, given the fact Date doesn't have timezone and I am not really certain what anyone will use the Date Object instead of a the seconds since UTC epoch :thinking: I am somewhat for returning it as a number to be honest

imiric replied:
You're right, fixed in f32a4d05.

na-- commented:
> List of problems that the original issue intended to solve:
> ....

@MStoykov, the original issue was just a catch-all for all of the somewhat similar "lack of information" problems that users experienced. In my mind, this PR was never meant to address _all_ of them, just to be at least an MVP for the trickiest and/or most useful ones first. After we merge this, I'm going to create a new issue with the remaining unsolved problems and we'll chip away at them in future PRs. It has grown to more than an MVP, and it addresses some things which are not necessarily problems or needs for anyone, so I am not completely opposed to pruning a few more things from it, but probably not everything you suggested...

> If we keep this, we either need to rename the fields to better show what they are or the method to `getInstanceStats`/`getTestIntanceStats`

Yeah, I was going to bring this up as well, the `getTestStats()` names and/or `getTestStats()` itself probably need a rename. We can discuss this and all of the names in Slack. We probably should have a consistent and logical naming scheme that makes clear where everything falls under (whole test (global) vs per instance, in scenario or out of it, etc.)

> `duration` - is the time the test has ran so far as a string, probably better to be float. 

I agree, already brought this up here in https://github.com/k6io/k6/pull/1863#discussion_r653507876

> progress - shows how far the executor has come as a float between 0 and 1. where 1 is completed. This is based on the max duration and is in some ways `time.now() - startTime/scenarioDuration` (this one isn't provided). No idea what that is useful for, and given that adding `scenarioDuration` will probably be way easier I wonder if we shouldn't add that instead

It sometimes depends on the number of iterations, it's not always the duration. I am not sure if users would be able to use it in a script, but might be useful for someone running with `--quiet` and wanting to log some progress occasionally :man_shrugging: 

> On this topic I am still of the opinion (now even more so) that the API is bad and it should look like :
> 
> 
> ```
> import exec from "k6/execution"; // exec might be changed
> 
> export default function() {
> 	exec.scenario.iteration // instead of  exec.getSecnarioStats().iterationGlobal
> 	exec.vu.id // instead of exec.getVUStats().idGlobal
> }
> ```

I disagree, I don't see what the problem with getter functions is, if anything, they add clarity to me. Consider that `k6/execution` will have more things in it, and they won't all be simply supplying information. For example, `abortTest()` (https://github.com/k6io/k6/pull/1920) should probably go in there. This seems quite nice to me:
```js
import execution from 'k6/execution';

export default function () {
    if (execution.getScenarioStats().iterationGlobal > 10000) {
        execution.abortTest();
    }
}
````

imiric commented:
@MStoykov @na-- [I extracted the JS extension](https://github.com/k6io/xk6-execution/pull/1) and cleaned this up a bit. Let me know if there are any leftover blockers here.

Do we need more tests? Most of them were written in JS and copied to the extension repo as is, but will need some tweaking to make them run there. So I'm wondering if we should rewrite some in Go to cover some scenarios here.

na-- commented:
I haven't looked at the new version of the PR yet, but regarding more tests, I think we probably don't need to add or convert any more tests now, unless it's something very quick... :man_shrugging:  All of the old k6 tests should verify that we haven't broken any old k6 behavior with the refactoring (:crossed_fingers:), and we always intend to bring the extension back into k6, so it will come with all of the tests then,

MStoykov commented on lib/testutils/minirunner/minirunner.go:
```diff
@@ -124,34 +131,58 @@ func (r *MiniRunner) HandleSummary(ctx context.Context, s *lib.Summary) (map[str
 	busy chan struct{}
+
+	scenarioName              string
+	getNexIterations          func() (uint64, uint64)
```
nit:typo 

```suggestion
	getNextIterations          func() (uint64, uint64)
```

imiric replied:
Resolved in a recent force-push.

MStoykov reviewed (approved):
LGTM!

I still think something should be dropped (progress in particular), but that can be discussed along the API in the extension

na-- commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -175,11 +175,23 @@ func (varc RampingArrivalRateConfig) HasWork(et *lib.ExecutionTuple) bool {
 // Make sure we implement the lib.Executor interface.
 var _ lib.Executor = &RampingArrivalRate{}
 
+// Init values needed for the execution
+func (varr *RampingArrivalRate) Init(ctx context.Context) error {
+	// err should always be nil, because Init() won't be called for executors
+	// with no work, as determined by their config's HasWork() method.
+	et, err := varr.BaseExecutor.executionState.ExecutionTuple.GetNewExecutionTupleFromValue(varr.config.MaxVUs.Int64)
+	varr.et = et
+	varr.iterSegIndex = lib.NewSegmentedIndex(et)
+
+	return err //nolint: wrapcheck
+}
```
Hmm this `Init()` makes sense, I think, but I wonder why we didn't have this re-segmentation before? Was that a bug we have to mention in the release notes separately?

imiric replied:
Not sure, it seems to segment properly to me on `master`, but maybe I'm not reproducing it.

If it was a minor bug we should mention it, sure.

MStoykov replied:
Apparently, I didn't do part of https://github.com/k6io/k6/issues/1308 :sweat: for this executor.

With 
```javascript
export let options = {
        scenarios: {
                contacts: {
                        executor: 'ramping-arrival-rate',
                        startRate: 5,
                        timeUnit: '1s',
                        preAllocatedVUs: 2,
                        maxVUs: 3,
                        stages:[ { target: 5, duration: "3s"}],
                },
        },
};
export default function () {
    console.log(__VU, __ITER);
}
```
(Notice the `maxVUs: 3` and preallocatedVUs: '2', it important and they both can't be proportionally split given a segment)
You can clearly see that with master and `--execution-segment '0:1/2'` it does  7 iterations (there re in total 15 = 3s * 5 iter/s).
With the changes from this branch it does 10 or 2/3 of the iterations as when you scale based on the maxVUs it gets 2 maxVUs out of 3. 
So ... good thing we caught and we should mention it. I doubt anyone actually got bitten by this or even if all that many people experienced it unknowingly, but it's definitely still a bug

na-- replied:
@imiric can you please add a link to or a copy of :arrow_up: in the PR description, so when we link it in the v0.33.0 release notes, it's clear that it also fixes it?

na-- commented on lib/executor/constant_vus.go:
```diff
@@ -176,15 +176,26 @@ func (clv ConstantVUs) Run(parentCtx context.Context, out chan<- stats.SampleCon
 		activeVUs.Done()
 	}
+
+	// Channel for synchronizing scenario-specific iteration increments
```
This comment seems to be for an earlier version of the PR?

imiric replied:
:facepalm: Merge gone awry... It was also in `per_vu_iterations.go`. Fixed in latest push.

Some of the intermediate commits are definitely broken, but :shrug:

na-- reviewed (approved):
LGTM besides the comment. And I am still not sure if this change doesn't have any side-effects - it looks correct, but as I mentioned in the comment, does that mean we had a minor bug before? :man_shrugging:

na-- reviewed (approved):
:shipit:

MStoykov reviewed (approved):
LGTM :clap:
